pkg/kubefed2/federate.go

```diff
@@ -0,0 +1,292 @@
+	opts := &federateType{}
+
+	cmd := &cobra.Command{
+		Use:     "federate --group=GROUP --version=VERSION --kind=KIND --host-cluster-context=HOST_CONTEXT",
+		Short:   "Enable propagation of a Kubernetes API type",
+		Long:    federate_long,
+		Example: federate_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := opts.Complete(args)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+
+			err = opts.Run(cmdOut, config)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+		},
+	}
+
+	flags := cmd.Flags()
+	opts.CommonBind(flags)
+	opts.Bind(flags)
+
+	return cmd
+}
+
+// Complete ensures that options are valid and marshals them if necessary.
+func (j *federateType) Complete(args []string) error {
+	if len(j.group) == 0 && j.version != "v1" {
+		return fmt.Errorf("--group is a mandatory parameter for non-core resources ")
+	}
+	if len(j.version) == 0 {
+		return fmt.Errorf("--version is a mandatory parameter")
+	}
+	if len(j.kind) == 0 {
+		return fmt.Errorf("--kind is a mandatory parameter")
+	}
+	if len(j.templateGroup) == 0 {
+		return fmt.Errorf("--template-group is a mandatory parameter ")
+	}
+	if len(j.templateVersion) == 0 {
+		return fmt.Errorf("--template-version is a mandatory parameter")
+	}
+	if j.rawComparisonField == string(apicommon.ResourceVersionField) ||
+		j.rawComparisonField == string(apicommon.GenerationField) {
+		j.comparisonField = apicommon.VersionComparisonField(j.rawComparisonField)
+	} else {
+		return fmt.Errorf("--comparison-field must be %q or %q",
+			apicommon.ResourceVersionField, apicommon.GenerationField,
+		)
+	}
+	if len(j.pluralName) == 0 {
+		j.pluralName = fedv1a1.PluralName(j.kind)
+	}
+
+	return nil
+}
+
+// Run is the implementation of the `federate` command.
+func (j *federateType) Run(cmdOut io.Writer, config util.FedConfig) error {
+	hostConfig, err := config.HostConfig(j.HostClusterContext, j.Kubeconfig)
+	if err != nil {
+		return fmt.Errorf("Failed to get host cluster config: %v", err)
+	}
+
+	fedClient, err := util.FedClientset(hostConfig)
+	if err != nil {
+		return fmt.Errorf("Failed to get federation clientset: %v", err)
+	}
+
+	crdClient, err := apiextv1b1client.NewForConfig(hostConfig)
+	if err != nil {
+		return fmt.Errorf("Failed to create crd clientset: %v", err)
+	}
+
+	apiResource := metav1.APIResource{
+		Group:      j.group,
+		Version:    j.version,
+		Kind:       j.kind,
+		Name:       j.pluralName,
+		Namespaced: j.namespaced,
+	}
+
+	// TODO(marun) Perform this check with the discovery client.
+	targetClient, err := controllerutil.NewResourceClientFromConfig(hostConfig, &apiResource)
+	if err != nil {
+		return fmt.Errorf("Error creating client for target resource %q: %v", apiResource.Kind, err)
+	}
+	_, err = targetClient.Resources(j.FederationNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Unable to verify that %q is a valid target type: %v", apiResource.Kind, err)
```

Comment:
I am using `kubefed2 federate` to federate resource `MutatingWebhookConfiguration` and I get the following error:
```
$ ./bin/kubefed2 federate --namespaced=true --group=admissionregistration.k8s.io --version=v1beta1 --kind=MutatingWebhookConfiguration
F1019 19:08:11.706428   25028 federate.go:113] error: Unable to verify that "MutatingWebhookConfiguration" is a valid target type: the server could not find the requested resource
```

The proper admission controllers are enabled on kube-apiserver and the necessary api-version is available:
```
root     30494 30466  4 Oct18 ?        00:54:38 kube-apiserver --authorization-mode=Node,RBAC --enable-admission-plugins=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota --advertise-address=10.138.0.2 --allow-privileged=true --client-ca-file=/var/lib/localkube/certs/ca.crt --disable-admission-plugins=PersistentVolumeLabel --enable-bootstrap-token-auth=true --etcd-cafile=/var/lib/localkube/certs/etcd/ca.crt --etcd-certfile=/var/lib/localkube/certs/apiserver-etcd-client.crt --etcd-keyfile=/var/lib/localkube/certs/apiserver-etcd-client.key --etcd-servers=https://127.0.0.1:2379 --insecure-port=0 --kubelet-client-certificate=/var/lib/localkube/certs/apiserver-kubelet-client.crt --kubelet-client-key=/var/lib/localkube/certs/apiserver-kubelet-client.key --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname --proxy-client-cert-file=/var/lib/localkube/certs/front-proxy-client.crt --proxy-client-key-file=/var/lib/localkube/certs/front-proxy-client.key --requestheader-allowed-names=front-proxy-client --requestheader-client-ca-file=/var/lib/localkube/certs/front-proxy-ca.crt --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-group-headers=X-Remote-Group --requestheader-username-headers=X-Remote-User --secure-port=8443 --service-account-key-file=/var/lib/localkube/certs/sa.pub --service-cluster-ip-range=10.96.0.0/12 --tls-cert-file=/var/lib/localkube/certs/apiserver.crt --tls-private-key-file=/var/lib/localkube/certs/apiserver.key

$ kubectl api-versions | grep admissionregistration.k8s.io/v1beta1
admissionregistration.k8s.io/v1beta1
```

Reply:
Consider specifying `--namespaced=false`

```
NAME                                     SHORTNAMES   APIGROUP                            NAMESPACED   KIND
mutatingwebhookconfigurations                         admissionregistration.k8s.io        false        MutatingWebhookConfiguration
```

Reply:
The requirement to specify `--namespaced` is temporary.  I'm working on looking up as much as possible from the host cluster to minimize what can go wrong.

Reply:
That did the trick. Sorry for not trying that before pinging you. It looks like the Istio manifest incorrectly states a ns for the resource:
```
# Source: istio/charts/sidecarInjectorWebhook/templates/mutatingwebhook.yaml
apiVersion: admissionregistration.k8s.io/v1beta1
kind: MutatingWebhookConfiguration
metadata:
  name: istio-sidecar-injector
  namespace: istio-system
<SNIP>
```